[Merged by Bors] - feat(Combinatorics/SimpleGraph): prove the minimal-degree version of the Erdős-Stone theorem

mitchell-horner opened the pull request:
Proves the minimal degree-version of the Erdős-Stone theorem:

If `G` has a minimal degree of at least `(1 - 1 / r + o(1)) * card V`, then `G` contains a copy of a `completeEquipartiteGraph` in `r + 1` parts each of size `t`.

The double-counting construction from the proof is available in `namespace ErdosStone`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

- [x] depends on: #25843
- [x] depends on: #27597
- [x] depends on: #27599
- [x] depends on: #28443
- [x] depends on: #28445
- [x] depends on: #28446
- [x] depends on: #28447

This is the first of several pull requests towards the full Erdős-Stone(-Simonovits) theorem, hence the name of the file.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

mitchell-horner committed:
Add Bipartite section

mitchell-horner committed:
Impl `between_le`

mitchell-horner committed:
Merge branch 'master' into mitchell-horner/between

mitchell-horner committed:
Merge branch 'master' into mitchell-horner/between

mitchell-horner committed:
Impl `between_isBipartite`

mitchell-horner committed:
Merge branch 'master' into mitchell-horner/between

mitchell-horner committed:
Merge branch 'master' into mitchell-horner/between

mitchell-horner committed:
Impl DecidablePred

mitchell-horner committed:
Impl DecidablePred

mitchell-horner committed:
Impl `completeEquipartiteGraph`

mitchell-horner committed:
Impl `completeEquipartiteGraph`

mitchell-horner committed:
Remove Fintype assumption

mitchell-horner committed:
Remove Fintype assumption

mitchell-horner committed:
Remove `simp` tag

mitchell-horner committed:
Remove `simp` tag

mitchell-horner committed:
Impl `completeEquipartiteSubgraph`

mitchell-horner committed:
Update docstrings

mitchell-horner committed:
Fix formatting

mitchell-horner committed:
Fix `completeEquipartiteGraph.completeMultipartiteGraph`

mitchell-horner committed:
Refactor `isContained_completeEquipartiteGraph_of_colorable`

mitchell-horner committed:
Impl `completeEquipartiteGraph.turanGraph`

mitchell-horner committed:
Update docstrings

mitchell-horner committed:
Fix formatting

mitchell-horner committed:
Move `section Coloring`

mitchell-horner committed:
Merge branch 'master' into complete-equipartite

mitchell-horner committed:
Impl `filter_disjiUnion`

mitchell-horner committed:
Impl `one_sub_one_div_cast_*` theorems

mitchell-horner committed:
Move `card_edgeFinset_eq`

mitchell-horner committed:
Impl `Colorable.map`

mitchell-horner committed:
Impl `chromaticNumber_eq_iff_colorable_not_colorable`

mitchell-horner committed:
Impl `free_of_colorable`

mitchell-horner committed:
Add imports

mitchell-horner committed:
Implicit `G`

mitchell-horner committed:
Unused imports

mitchell-horner committed:
Impl `edgeSet_map_eq_image`

mitchell-horner committed:
Impl `edgeFinset_map_eq_map`

mitchell-horner committed:
Explicit `G`

mitchell-horner committed:
Refactoring

mitchell-horner committed:
Drop `private` + Refactor

mitchell-horner committed:
Merge branch 'master' into disjiUnion-results

mitchell-horner committed:
Refactor `completeEquipartiteGraph.completeMultipartiteGraph`

mitchell-horner committed:
Oneline `completeEquipartiteGraph.completeMultipartiteGraph`

mitchell-horner committed:
Refactor `Colorable.map`

mitchell-horner committed:
Renaming

mitchell-horner committed:
`f` as hypothesis

mitchell-horner committed:
Tag `edgeFinset_map` with `simp`

mitchell-horner committed:
Merge branch 'master' into complete-equipartite

mitchell-horner committed:
Merge branch 'master' into map-results

mitchell-horner committed:
Merge branch 'master' into coloring-results

mitchell-horner committed:
Merge branch 'master' into erdos-stone-1

mitchell-horner committed:
Merge remote-tracking branch 'origin/mitchell-horner/between' into erdos-stone-1

mitchell-horner committed:
Merge branch 'disjiUnion-results' into erdos-stone-1

mitchell-horner committed:
Merge branch 'one_sub-results' into erdos-stone-1

mitchell-horner committed:
Merge branch 'map-results' into erdos-stone-1

mitchell-horner committed:
Merge branch 'coloring-results' into erdos-stone-1

mitchell-horner committed:
Merge branch 'complete-equipartite' into erdos-stone-1

mitchell-horner committed:
Add ErdosStoneSimonovits.lean

mitchell-horner committed:
Fix formatting

mitchell-horner committed:
Merge branch 'master' into erdos-stone-1

mitchell-horner committed:
Fix imports

mitchell-horner committed:
Remove `Asymptotics`

mitchell-horner committed:
Merge branch 'master' into erdos-stone-1

mitchell-horner committed:
Update ErdosStoneSimonovits.lean

YaelDillies reviewed (commented):
Why do you have both an `ErdosStone` section and an `ErdosStone` namespace?

mitchell-horner committed:
Merge branch 'master' into erdos-stone-1

mitchell-horner committed:
Address comments

mitchell-horner commented on Mathlib/Combinatorics/SimpleGraph/Extremal/ErdosStoneSimonovits.lean:
```diff
@@ -0,0 +1,322 @@
+  minimal degree version of the **Erdős-Stone theorem** for simple graphs.
+-/
+
+@[expose] public section
```
I've put `public` in front of 
```
public theorem eventually_completeEquipartiteGraph_isContained_of_minDegree
```

Is this correct? :sweat_smile:

mitchell-horner commented on Mathlib/Combinatorics/SimpleGraph/Extremal/ErdosStoneSimonovits.lean:
```diff
@@ -0,0 +1,322 @@
+    Finset.prod_congr rfl fun p hp ↦ show (#p).choose t = t'.choose t by rw [K.card_mem_parts hp],
+    prod_const, K.card_parts.resolve_right ht'_pos.ne']
+  exact_mod_cast le_of_mul_le_mul_right (mul_le_card_filter_mul K hr_pos ht'_pos hδ (mod_cast hN))
+    (sub_pos_of_lt <| mod_cast ht_lt_t' : 0 < (t' - t : ℝ))
```
It is!

mitchell-horner commented:
> Why do you have both an `ErdosStone` section and an `ErdosStone` namespace?

@YaelDillies The idea was to put all of the private theorems for the `ErdosStone` section in an `ErdosStone` namespace for readability. (I left them without private because I was unsure if it would be useful to re-use the construction elsewhere. They should all be private now.)

I have a `section ErdosStoneSimonovits` and a `namespace ErdosStoneSimonovits` further down the file (in a dependent pull request) with the same intention.

(I also do the same thing [over here](https://github.com/leanprover-community/mathlib4/pull/25841) for `section KovariSosTuran`. I will apply the same changes to section/namespace/public/private in that pull request.)

I am happy to just put `public` in front of the few results that actually need it for this file.

EDIT: Sorry if this setup is a bit odd. I am a little unfamiliar with the visibility modifiers and took a stab what I thought looked tidy.

YaelDillies commented:
IME what works best in this sort of situation is indeed to mark as `public` only the few declarations that should be public, and not use `public section`

b-mehta commented:
> I am a little unfamiliar with the visibility modifiers

Very briefly, there's public/private and they're opposites, public lemmas are usable downstream and private ones aren't. And you can have private or public section (the default is private, but mathlib uses mostly public), and inside a section you can flip the specific one. There's also @[expose] but I think that's not relevant here

robin-carlier commented:
@b-mehta coming from the triage team, do you think you can have a look at this? It seems the comment about visibility has been addressed.

YaelDillies commented:
We're waiting on Mitchell to implement the visibility changes

mitchell-horner commented:
> We're waiting on Mitchell to implement the visibility changes

@robin-carlier is correct. I addressed the comments in the commit with message `Address comments`.

@YaelDillies is there anything I missed?

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Extremal/ErdosStoneSimonovits.lean:
```diff
@@ -0,0 +1,329 @@
+/-- `#ErdosStone.filter` is arbitrarily large with respect to `n`.
+
+This is an auxiliary definition for the **Erdős-Stone theorem**. -/
+theorem mul_le_card_filter_mul (hr_pos : 0 < r) (ht'_pos : 0 < t')
```
This is not a definition. Same below

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Extremal/ErdosStoneSimonovits.lean:
```diff
@@ -0,0 +1,329 @@
+-/
+
+
+open Filter Finset Fintype Real Topology
```
```suggestion
open Filter Finset Fintype Real Topology
```

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Extremal/ErdosStoneSimonovits.lean:
```diff
@@ -0,0 +1,329 @@
+subgraph, in `r` parts each of size `t'`, adjacent to at least `t` vertices in each part of the
+complete equipartite subgraph.
+
+This is an auxiliary definition for the **Erdős-Stone theorem**. -/
```
No need to add emphasis on the lemma docstrings
```suggestion
This is an auxiliary definition for the Erdős-Stone theorem. -/
```

mitchell-horner committed:
Update docstrings + Whitespace

YaelDillies commented on Mathlib/Combinatorics/SimpleGraph/Extremal/ErdosStoneSimonovits.lean:
```diff
@@ -120,7 +119,7 @@ lemma card_edgeFinset_between_verts_le (hr_pos : 0 < r) (ht'_pos : 0 < t') :
 /-- `#ErdosStone.filter` is arbitrarily large with respect to `n`.
 
-This is an auxiliary definition for the **Erdős-Stone theorem**. -/
+This is an auxiliary theorem for the Erdős-Stone theorem. -/
```
```suggestion
This is an auxiliary lemma for the Erdős-Stone theorem. -/
```
surely?

mitchell-horner committed:
Replace theorem with lemma

YaelDillies reviewed (approved):
Thanks! 🚀

maintainer merge

mitchell-horner committed:
Merge branch 'master' into erdos-stone-1

mitchell-horner committed:
Update ErdosStoneSimonovits.lean

b-mehta reviewed (commented):
Thank you!

bors merge
